### DISCUSSION TITLE
Why does vLLM use a custom all-reduce method?

### DISCUSSION THREAD

**Discussion by SamKG** [vllm-project/vllm#6159] [Q&A]
Hello,
Was hoping someone could help shed some light on this - 
why does vLLM choose to use a custom all-reduce method? Is there a benefit to doing this over just using the NCCL APIs?

**Comment by simon-mo** [comment]
See perf result here https://github.com/vllm-project/vllm/pull/2192. In certain cases, the custom topology drastically boosts performance compared to nccl's implementation. vLLM still uses nccl in majority of cases.

**Reply by SamKG** [reply]
thanks, that helps a ton!!

**Reply by SamKG** [reply]
@simon-mo  Just wanted to quickly clarify something: I've observed strange scaling in performance for the 1-stage all-reduce kernels (both nccl and for vllm):

![image](https://github.com/vllm-project/vllm/assets/15336495/f007998c-9e39-448a-9962-2a3200a6d3ed)

For small problem sizes, it actually performs *slower* than expected (this is the region where 1-stage reduce is being used)

Theoretically, we should expect scalings that look (roughly) like the below:
![image](https://github.com/vllm-project/vllm/assets/15336495/6e16aef6-dd31-49b6-adcf-72aadafb0baa)




Is this expected?

**Reply by SamKG** [reply]
cc @hanzhi713

**Comment by cduk** [comment]
In a set-up where 4 GPUs are connected by PCIe, but each pair of GPUs are connected by NVLink (112 GB/s bi-directional). Is there a way to specify a reduction first on each pairwise bound set of GPUs before reducing across the slower PCIe link?
